lib/tree.rb

```diff
@@ -16,48 +16,117 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(1)
+  # Space Complexity: O(n)
+  def add(key, value = nil)
```

Comment:
👍 The time complexity is O(n) if the tree is unbalanced and O(log n) if the tree is balanced.

Same with the space complexity due to the system stack.